lib/pbench/agent/tool_data_sink.py

```diff
@@ -1609,71 +1831,225 @@ def put_document(self, data_ctx, hostname):
+    with open(f"{PROG}.out", "w") as sofp, open(
+        f"{PROG}.err", "w"
+    ) as sefp, DaemonContext(
+        stdout=sofp,
+        stderr=sefp,
+        working_directory=os.getcwd(),
+        umask=0o022,
+        pidfile=pfctx,
+    ):
+        logger = get_logger(PROG, daemon=True)
+
+        # We have to re-open the connection to the redis server now that we
+        # are "daemonized".
+        logger.debug("re-constructing Redis server object")
+        try:
+            redis_server = redis.Redis(host=redis_host, port=redis_port, db=0)
+        except Exception as e:
+            logger.error(
+                "Unable to construct Redis server object, %s:%s: %s",
+                redis_host,
+                redis_port,
+                e,
+            )
+            return 7
+        else:
+            logger.debug("reconstructed Redis server object")
+        return driver(
+            PROG,
+            redis_server,
+            redis_host,
+            redis_port,
+            pbench_bin,
+            pbench_run,
+            hostname,
+            tar_path,
+            cp_path,
+            param_key,
+            params,
+            optional_md,
+            logger=logger,
+        )
+
+
+def main(argv):
+    _prog = Path(argv[0])
+    PROG = _prog.name
+    # The Tool Data Sink executable is in:
+    #   ${pbench_bin}/util-scripts/tool-meister/pbench-tool-data-sink
+    # So .parent at each level is:
+    #   _prog       ${pbench_bin}/util-scripts/tool-meister/pbench-tool-data-sink
+    #     .parent   ${pbench_bin}/util-scripts/tool-meister
+    #     .parent   ${pbench_bin}/util-scripts
+    #     .parent   ${pbench_bin}
+    pbench_bin = _prog.parent.parent.parent
+
     try:
         redis_host = argv[1]
         redis_port = argv[2]
         param_key = argv[3]
     except IndexError as e:
-        logger.error("Invalid arguments: %s", e)
+        print(f"{PROG}: Invalid arguments: {e}", file=sys.stderr)
         return 1
+    else:
+        if not redis_host or not redis_port or not param_key:
+            print(f"{PROG}: Invalid arguments: {argv!r}", file=sys.stderr)
+            return 1
+    try:
+        daemonize = argv[4]
+    except IndexError:
+        daemonize = "no"
+    else:
+        if not daemonize:
+            daemonize = "no"
 
-    global tar_path
     tar_path = find_executable("tar")
     if tar_path is None:
-        logger.error("External 'tar' executable not found")
+        print("External 'tar' executable not found", file=sys.stderr)
         return 2
 
-    global cp_path
     cp_path = find_executable("cp")
     if cp_path is None:
-        logger.error("External 'cp' executable not found")
+        print("External 'cp' executable not found", file=sys.stderr)
         return 2
 
+    try:
+        pbench_run = os.environ["pbench_run"]
+    except KeyError:
+        print(
+            "Unable to fetch pbench_run environment variable", file=sys.stderr,
+        )
+        return 3
```

Comment:
I think we need to set up a 12-step program to get us off of using exceptions.  How about this, instead?
```
    if "pbench_run" not in os.environ:
        print(
            "Unable to fetch pbench_run environment variable", file=sys.stderr,
        )
        return 3
    pbench_run = os.environ["pbench_run"]
```
or, perhaps better,
```
    pbench_run = os.environ.get("pbench_run")
    if pbench_run is None:
        print(
            "Unable to fetch pbench_run environment variable", file=sys.stderr,
        )
        return 3
```

Reply:
See https://stackoverflow.com/questions/7604636/better-to-try-something-and-catch-the-exception-or-test-if-its-possible-first.

Reply:
I may have to take back my comment about the 12-step program and getting off of exceptions:  while I still think using `try` blocks makes the code harder to read, they definitely seem to make it perform _better_, which is counterintuitive, given what I know about _implementing_ exception handling....